File: python/src/train/layers/towers.py
from __future__ import annotations

"""多任务塔：TaskTower、MultiTaskTower、任务关系推导。"""
from dataclasses import dataclass, field
from enum import Enum, auto

import torch
import torch.nn as nn
import torch.nn.functional as F

from train.core.model_output import ModelOutput


class Activation(Enum):
    RELU = auto()
    SIGMOID = auto()
    SWISH = auto()
    GELU = auto()
    NONE = auto()

    def apply(self, x: torch.Tensor) -> torch.Tensor:
        """Apply activation function to tensor."""
        m = {
            self.RELU: F.relu,
            self.SIGMOID: torch.sigmoid,
            self.SWISH: lambda x: x * torch.sigmoid(x),
            self.GELU: F.gelu,
            self.NONE: lambda x: x,
        }
        return m[self](x)

    @classmethod
    def from_str(cls, s: str) -> Activation:
        """Parse activation from string: relu, sigmoid, swish, gelu, none."""
        return {
            "relu": cls.RELU,
            "sigmoid": cls.SIGMOID,
            "swish": cls.SWISH,
            "gelu": cls.GELU,
            "none": cls.NONE,
        }[s.lower()]


@dataclass
class TowerConfig:
    name: str
    hidden_dims: list[int] = field(default_factory=list)
    output_dim: int = 1
    activation: Activation = Activation.RELU
    output_kind: str = "binary_logit"


class TaskTower(nn.Module):
    """Single-task MLP tower. Naming matches Candle vb.pp paths."""

    def __init__(self, config: TowerConfig, input_dim: int) -> None:
        """Build tower from TowerConfig with hidden.{i} and output.{n} naming."""
        super().__init__()
        self.name = config.name
        self.output_kind = config.output_kind
        self._act = config.activation
        self.hidden = nn.ModuleDict()
        in_dim = input_dim
        for i, h in enumerate(config.hidden_dims):
            self.hidden[str(i)] = nn.Linear(in_dim, h)
            in_dim = h
        n = len(config.hidden_dims)
        self.output = nn.ModuleDict()
        self.output[str(n)] = nn.Linear(in_dim, config.output_dim)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        """Forward: hidden layers with activation, output layer without."""
        for i in range(len(self.hidden)):
            x = self._act.apply(self.hidden[str(i)](x))
        return self.output[str(len(self.hidden))](x)


@dataclass
class TaskRelation:
    target: str
    sources: list[str]
    op: str


@dataclass
class MultiTaskConfig:
    towers: list[TowerConfig]
    relations: list[TaskRelation] = field(default_factory=list)


class MultiTaskTower(nn.Module):
    """Multi-task tower manager with optional task relation derivation."""

    def __init__(self, config: MultiTaskConfig, input_dim: int) -> None:
        """Build towers from config; each registered by name matching Candle vb.pp paths."""
        super().__init__()
        for tc in config.towers:
            setattr(self, tc.name, TaskTower(tc, input_dim))
        self._tower_names = [tc.name for tc in config.towers]
        self._relations = config.relations
        self._relation_names = [rel.target for rel in config.relations]

    def forward(self, shared: torch.Tensor) -> ModelOutput:
        """Run all towers, then apply task relations (sigmoid before op)."""
        outputs = ModelOutput()
        for name in self._tower_names:
            tower = getattr(self, name)
            outputs.insert(name, tower(shared), tower.output_kind)
        for rel in self._relations:
            probs = {s: _relation_probability(rel, outputs, s) for s in rel.sources}
            left = probs[rel.sources[0]]
            right = probs[rel.sources[1]]
            if rel.op == "multiply":
                outputs.insert_probability(rel.target, left * right)
            elif rel.op == "add":
                outputs.insert_probability(rel.target, left + right)
            elif rel.op == "subtract":
                outputs.insert_probability(rel.target, left - right)
            elif rel.op == "divide":
                outputs.insert_probability(rel.target, left / (right + 1e-8))
            else:
                raise ValueError(f"unsupported task relation op: {rel.op}")
        return outputs

    @property
    def relation_names(self) -> list[str]:
        return list(self._relation_names)


def _relation_probability(rel: TaskRelation, outputs: ModelOutput, source: str) -> torch.Tensor:
    output = outputs.get(source)
    if output is None:
        raise ValueError(f"Relation '{rel.target}' source '{source}' is missing")
    if output.kind == "binary_logit":
        return torch.sigmoid(output.tensor)
    if output.kind == "probability":
        return output.tensor
    raise ValueError(
        f"Relation '{rel.target}' source '{source}' must be binary_logit or probability, "
        f"got {output.kind}"
    )

File: src/layers/towers.rs
//! 多任务预测塔：TaskTower、MultiTaskTower、任务关系推导。
use candle_core::{Result, Tensor};
use candle_nn::{linear, Linear, Module, VarBuilder};
use serde::{Deserialize, Serialize};
use std::collections::HashSet;

use crate::models::{ModelOutput, OutputKind};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
/// 激活函数类型。
pub enum Activation {
    /// ReLU 激活。
    Relu,
    /// Sigmoid 激活。
    Sigmoid,
    /// Swish 激活。
    Swish,
    /// GELU 激活。
    Gelu,
    /// 不使用激活函数。
    #[serde(rename = "none")]
    None_,
}
impl Default for Activation {
    fn default() -> Self {
        Activation::Relu
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
/// 单任务塔配置：名称、隐藏层维度、输出维度、激活函数。
pub struct TowerConfig {
    /// 任务塔名称，也是输出名称。
    pub name: String,
    /// 隐藏层维度列表。
    #[serde(default)]
    pub hidden_dims: Vec<usize>,
    /// 输出维度。
    pub output_dim: usize,
    /// 隐藏层激活函数。
    #[serde(default)]
    pub activation: Activation,
    /// 输出语义类型。
    #[serde(default = "default_output_kind")]
    pub output_kind: OutputKind,
}

fn default_output_kind() -> OutputKind {
    OutputKind::BinaryLogit
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
/// 任务间概率关系运算类型。
pub enum RelationOp {
    /// 源概率相乘。
    Multiply,
    /// 源概率相加。
    Add,
    /// 两个源概率相减。
    Subtract,
    /// 两个源概率相除。
    Divide,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
/// 任务关系定义：目标任务由源任务经运算推导。
pub struct TaskRelation {
    /// 派生输出名称。
    pub target: String,
    /// 参与关系计算的源输出名称。
    pub sources: Vec<String>,
    /// 概率关系运算。
    pub op: RelationOp,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
/// 多任务塔完整配置：塔列表 + 关系推导列表。
pub struct MultiTaskConfig {
    /// 独立任务塔配置列表。
    pub towers: Vec<TowerConfig>,
    /// 派生概率关系配置列表。
    #[serde(default)]
    pub relations: Vec<TaskRelation>,
}

/// 单任务预测塔。
///
/// 多层 MLP + 层间激活，输出 logits（末层无激活）。
/// 参数命名: `hidden.{i}` 和 `output.{num_hidden}` 匹配 Candle 路径。
pub struct TaskTower {
    name: String,
    layers: Vec<Linear>,
    activation: Activation,
    output_kind: OutputKind,
}

impl TaskTower {
    /// 根据配置构造塔。`input_dim` 为共享底层输出维度。
    pub fn new(config: &TowerConfig, input_dim: usize, vb: VarBuilder) -> Result<Self> {
        let mut layers = Vec::new();
        let mut in_dim = input_dim;
        let num_hidden = config.hidden_dims.len();
        for (i, &h_dim) in config.hidden_dims.iter().enumerate() {
            layers.push(linear(in_dim, h_dim, vb.pp(format!("hidden.{}", i)))?);
            in_dim = h_dim;
        }
        layers.push(linear(
            in_dim,
            config.output_dim,
            vb.pp(format!("output.{}", num_hidden)),
        )?);
        Ok(Self {
            name: config.name.clone(),
            layers,
            activation: config.activation.clone(),
            output_kind: config.output_kind,
        })
    }

    fn apply_activation(&self, x: &Tensor) -> Result<Tensor> {
        match self.activation {
            Activation::Relu => x.relu(),
            Activation::Sigmoid => candle_nn::ops::sigmoid(x),
            Activation::Swish => {
                let sig = candle_nn::ops::sigmoid(x)?;
                x.mul(&sig)
            }
            Activation::Gelu => x.gelu(),
            Activation::None_ => Ok(x.clone()),
        }
    }

    /// 前向: hidden → act → ... → output logits。
    /// 前向：各塔独立输出 → 关系推导（sigmoid 后运算）。
    /// 返回 `{task_name: logits}` 字典。
    pub fn forward(&self, x: &Tensor) -> Result<Tensor> {
        let mut out = x.clone();
        let n = self.layers.len();
        for (i, layer) in self.layers.iter().enumerate() {
            out = layer.forward(&out)?;
            if i < n - 1 {
                out = self.apply_activation(&out)?;
            }
        }
        Ok(out)
    }

    /// 返回该塔的任务名称。
    pub fn name(&self) -> &str {
        &self.name
    }

    /// 返回该塔输出的语义类型。
    pub fn output_kind(&self) -> OutputKind {
        self.output_kind
    }
}

impl Module for TaskTower {
    fn forward(&self, xs: &Tensor) -> Result<Tensor> {
        self.forward(xs)
    }
}

/// 多任务塔管理器。
///
/// 管理独立塔的前向传播和任务间概率关系推导（如 CTCVR = CTR × CVR）。
pub struct MultiTaskTower {
    towers: Vec<TaskTower>,
    relations: Vec<TaskRelation>,
    relation_targets: HashSet<String>,
}

impl MultiTaskTower {
    /// 根据配置构建多任务塔。
    pub fn new(config: &MultiTaskConfig, input_dim: usize, vb: VarBuilder) -> Result<Self> {
        let mut towers = Vec::new();
        for tc in &config.towers {
            let tower_vb = vb.pp(tc.name.clone());
            towers.push(TaskTower::new(tc, input_dim, tower_vb)?);
        }
        Ok(Self {
            towers,
            relations: config.relations.clone(),
            relation_targets: config
                .relations
                .iter()
                .map(|relation| relation.target.clone())
                .collect(),
        })
    }

    /// 返回所有任务名称（含推导任务）。
    pub fn task_names(&self) -> Vec<String> {
        let mut names: Vec<String> = self.towers.iter().map(|t| t.name().to_string()).collect();
        for r in &self.relations {
            names.push(r.target.clone());
        }
        names
    }

    /// 返回独立塔的数量（不含推导任务）。
    pub fn num_towers(&self) -> usize {
        self.towers.len()
    }

    /// 判断输出是否为关系推导出的概率。
    pub fn is_relation_output(&self, name: &str) -> bool {
        self.relation_targets.contains(name)
    }

    /// 前向: hidden → act → ... → output logits。
    /// 前向：各塔独立输出 → 关系推导（sigmoid 后运算）。
    /// 返回 `{task_name: logits}` 字典。
    pub fn forward(&self, shared_output: &Tensor) -> Result<ModelOutput> {
        let mut outputs = ModelOutput::new();
        for tower in &self.towers {
            outputs.insert(
                tower.name().to_string(),
                tower.forward(shared_output)?,
                tower.output_kind(),
            );
        }
        for rel in &self.relations {
            let derived = apply_relation(rel, &outputs)?;
            outputs.insert_probability(rel.target.clone(), derived);
        }
        Ok(outputs)
    }

    /// 返回各独立塔的 logits（不含推导任务）。
    pub fn tower_logits(&self, shared_output: &Tensor) -> Result<Vec<Tensor>> {
        self.towers
            .iter()
            .map(|t| t.forward(shared_output))
            .collect()
    }
}

/// 应用任务间概率关系推导（如 CTCVR = sigmoid(click) × sigmoid(cvr)）。
///
/// 从 `outputs` 中取出 `sources` 任务的 logits，经 sigmoid 转换后按 `op` 运算，
/// 返回推导目标的概率值。
pub fn apply_relation(rel: &TaskRelation, outputs: &ModelOutput) -> Result<Tensor> {
    if rel.sources.is_empty() {
        return Err(candle_core::Error::Msg(format!(
            "relation '{}' has no sources",
            rel.target
        )));
    }
    let get_prob = |name: &str| -> Result<Tensor> {
        let output = outputs
            .get(name)
            .ok_or_else(|| candle_core::Error::Msg(format!("task '{}' not found", name)))?;
        match output.kind {
            OutputKind::BinaryLogit => candle_nn::ops::sigmoid(&output.tensor),
            OutputKind::Probability => Ok(output.tensor.clone()),
            OutputKind::Regression | OutputKind::Score => Err(candle_core::Error::Msg(format!(
                "relation '{}' source '{}' must be binary_logit or probability, got {:?}",
                rel.target, name, output.kind
            ))),
        }
    };
    match rel.op {
        RelationOp::Multiply => {
            let mut result = get_prob(&rel.sources[0])?;
            for name in &rel.sources[1..] {
                result = result.mul(&get_prob(name)?)?;
            }
            Ok(result)
        }
        RelationOp::Add => {
            let mut result = get_prob(&rel.sources[0])?;
            for name in &rel.sources[1..] {
                result = result.broadcast_add(&get_prob(name)?)?;
            }
            Ok(result)
        }
        RelationOp::Subtract => {
            if rel.sources.len() != 2 {
                return Err(candle_core::Error::Msg(format!(
                    "relation '{}' subtract requires 2 sources",
                    rel.target
                )));
            }
            get_prob(&rel.sources[0])?.broadcast_sub(&get_prob(&rel.sources[1])?)
        }
        RelationOp::Divide => {
            if rel.sources.len() != 2 {
                return Err(candle_core::Error::Msg(format!(
                    "relation '{}' divide requires 2 sources",
                    rel.target
                )));
            }
            get_prob(&rel.sources[0])?.broadcast_div(&get_prob(&rel.sources[1])?)
        }
    }
}
